contracts/extension/plugin/Map.sol

```diff
@@ -3,61 +3,103 @@ pragma solidity ^0.8.0;
 
 import "../interface/plugin/IMap.sol";
 
-import "../Multicall.sol";
 import "../../openzeppelin-presets/utils/EnumerableSet.sol";
 
 /**
  *  TODO:
  *      - Remove OZ EnumerableSet external dependency.
  */
 
-contract Map is IMap, Multicall {
+abstract contract Map is IMap {
     using EnumerableSet for EnumerableSet.Bytes32Set;
 
     EnumerableSet.Bytes32Set private allSelectors;
 
-    mapping(address => EnumerableSet.Bytes32Set) private selectorsForExtension;
-    mapping(bytes4 => address) private extension;
+    mapping(address => EnumerableSet.Bytes32Set) private selectorsForPlugin;
+    mapping(bytes4 => Plugin) private pluginForSelector;
 
-    constructor(ExtensionMap[] memory _extensionsToRegister) {
-        uint256 len = _extensionsToRegister.length;
+    constructor(Plugin[] memory _pluginsToRegister) {
+        uint256 len = _pluginsToRegister.length;
         for (uint256 i = 0; i < len; i += 1) {
-            _setExtensions(_extensionsToRegister[i].selector, _extensionsToRegister[i].extension);
+            _setPlugin(_pluginsToRegister[i]);
         }
     }
 
-    function _setExtensions(bytes4 _selector, address _extension) internal {
-        require(allSelectors.add(bytes32(_selector)), "REGISTERED");
+    function setPlugin(Plugin memory _plugin) external {
+        require(_canSetPlugin(), "Not authorized");
 
-        extension[_selector] = _extension;
-        selectorsForExtension[_extension].add(bytes32(_selector));
+        _setPlugin(_plugin);
+    }
+
+    function updatePlugin(Plugin memory _plugin) external {
+        require(_canSetPlugin(), "Not authorized");
+
+        _updatePlugin(_plugin);
+    }
+
+    function removePlugin(bytes4 _selector) external {
+        require(_canSetPlugin(), "Not authorized");
+
+        _removePlugin(_selector);
+    }
+
+    function _setPlugin(Plugin memory _plugin) internal {
+        require(allSelectors.add(bytes32(_plugin.selector)), "REGISTERED");
+        require(_plugin.selector == bytes4(keccak256(abi.encodePacked(_plugin.functionString))), "Incorrect selector");
+
+        pluginForSelector[_plugin.selector] = _plugin;
+        selectorsForPlugin[_plugin.pluginAddress].add(bytes32(_plugin.selector));
 
-        emit ExtensionRegistered(_selector, _extension);
+        emit PluginRegistered(_plugin.selector, _plugin.pluginAddress);
     }
 
-    function getExtensionForFunction(bytes4 _selector) external view returns (address) {
-        address ext = extension[_selector];
-        require(ext != address(0), "No extension available for selector.");
+    function _updatePlugin(Plugin memory _plugin) internal {
+        address currentExtension = getPluginForFunction(_plugin.selector);
+        require(_plugin.selector == bytes4(keccak256(abi.encodePacked(_plugin.functionString))), "Incorrect selector");
 
-        return ext;
+        pluginForSelector[_plugin.selector] = _plugin;
+        selectorsForPlugin[_plugin.pluginAddress].add(bytes32(_plugin.selector));
```

Comment:
Put this `add` call in a `require` statement.

If `updatePlugin` is called twice with the same `Plugin` struct info, this `add` call will fail silently and the following `remove` call will remove the selector as from the `selectorsForPlugin` set, thus, tainting data / state of the contract.

Reply:
Or, you can reverse the order of the `add` and `remove` calls.